[PyTorch] Integrate Execution Graph Observer into PyTorch Profiler

louisfeng opened the pull request:
Test Plan:
```
buck build mode/dev-nosan caffe2/test:profiler --show-output
buck-out/gen/caffe2/test/profiler#binary.par test_profiler.TestExecutionGraph.test_execution_graph
```

Example output: P491658589

Differential Revision: D35342394

albanD commented on torch/autograd/__init__.py:
```diff
@@ -305,6 +305,8 @@ def variable(*args, **kwargs):
 from torch._C._autograd import (DeviceType, ProfilerActivity, ProfilerState, ProfilerConfig, ProfilerEvent,
                                 _enable_profiler_legacy, _disable_profiler_legacy, _profiler_enabled,
                                 _enable_record_function, _set_empty_test_observer, kineto_available,
+                                register_execution_graph_observer_once, enable_execution_graph_observer,
```
Do we really want to expose these as public APIs in `torch.autograd`? Shouldn't they be in the `torch.profiler`?

louisfeng replied:
Thanks @albanD, this diff is still WIP. I need to do a few more passes before it's ready for review. Will ping you when it's ready :).

louisfeng replied:
@albanD, the code is mostly ready for review. There were some windows build issues, I'm trying to resolve. Otherwise, it's ready for review.

albanD reviewed (commented):
Just checked the binding code where I think we can remove some unnecessary code.
I'll let @robieta review the rest of the profiler changes as he's the expert.

louisfeng commented on torch/profiler/profiler.py:
```diff
@@ -9,8 +9,14 @@
 import torch.autograd.profiler as prof
-from torch.autograd import ProfilerActivity, kineto_available
 from torch._C._autograd import _ExperimentalConfig
+from torch.autograd import (
```
Thanks, fixed.

louisfeng commented on torch/autograd/__init__.py:
```diff
@@ -313,7 +313,9 @@ def variable(*args, **kwargs):
                                 _record_function_with_args_enter, _record_function_with_args_exit,
                                 _supported_activities, _add_metadata_json, SavedTensor,
-                                _push_saved_tensors_default_hooks, _pop_saved_tensors_default_hooks)
+                                _push_saved_tensors_default_hooks, _pop_saved_tensors_default_hooks,
```
Thanks, fixed.

robieta reviewed (approved):
LGTM. Reviewed internally.

louisfeng committed:
[PyTorch] Integrate Execution Graph Observer into PyTorch Profiler (#75358)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/75358

Test Plan:
```
buck build mode/dev-nosan caffe2/test:profiler --show-output
buck-out/gen/caffe2/test/profiler#binary.par test_profiler.TestExecutionGraph.test_execution_graph
```

Example output: P502008384

Reviewed By: fengxizhou, robieta

Differential Revision: D35342394

Pulled By: louisfeng

fbshipit-source-id: 5f9bdd0a66fe40dbe815c87f06815987a9012c13
